### PR TITLE
Set Weighted changes to the ammo crates for weapons

### LINKED ISSUE #485: Changed Crate spawn for weapons to use selectRandomWeighted
I'm taking the below.

https://github.com/official-antistasi-community/A3-Antistasi/blob/cf1ada52adb95055d95e68971e0926713c050b74/A3-Antistasi/functions/Ammunition/fn_NATOCrate.sqf#L34-L106

And started to replace it with..
```SQF
private _weaponlootWeighting = [
	allRifles, 3.2,
	allHandguns, 1.2,
	allMachineGuns, 2.2,
	allShotguns, 0,
	allSMGs, 2.2,
	allSniperRifles, 0.6,
	allMissileLaunchers, 0.3,
	allRocketLaunchers, 0.3
];

// Little function to ensure the item isn't already unlocked.
private _fnc_pickRandomFromANotInB = {
	params ["_arrayA", "_arrayB"];
	private _choice = selectRandom _arrayA;
	private _foundValid = true;
	if (_choice in _arrayB) then {
		_foundValid = false;
		for "_i" from 0 to 9 do {
			_choice = selectRandom _arrayA;
			//We did it!
			if !(_choice in _arrayB) exitWith {
				_foundValid = true;
			};
		}
	};

	if (_foundValid) then {
		_choice;
	};
};

//Weapons Loot
[3, "Generating Weapons", "fn_NATOCrate"] call A3A_fnc_log;
for "_i" from 0 to floor random _crateWepTypeMax do {
	private _selection = selectRandomWeighted _weaponlootWeighting;
	[3, format ["Selected: %1", _selection],"fn_NATOCrate"] call A3A_fnc_log;
	private _loot =	 selectRandom _selection;
	[3, format ["%1 weapons chosen", _i],"fn_NATOCrate"] call A3A_fnc_log;
	[3, format ["Final weapon: %1", _loot],"fn_NATOCrate"] call A3A_fnc_log;

	if (isNil "_loot") then {
		[3, "No Weapons Left in Loot List Or Pick Random Failed","fn_NATOCrate"] call A3A_fnc_log;
	}
	else {
		_amount = floor random crateWepNumMax;
		_crate addWeaponWithAttachmentsCargoGlobal [[ _loot, "", "", "", [], [], ""], _amount];
		for "_i" from 0 to _amount do {
			_magazines = getArray (configFile / "CfgWeapons" / _loot / "magazines");
			[3, format ["Grabbing a %1 for %2", _magazines, _loot],"fn_NATOCrate"] call A3A_fnc_log;
			_magAmount = selectRandom [0,1,2];
			[3, format ["Spawning %1 magazines for %2", _magAmount, _loot],"fn_NATOCrate"] call A3A_fnc_log;
			_crate addMagazineCargoGlobal [_magazines select 0, _magAmount];
			[3, format ["Spawning %1 of %2", _amount, _loot],"fn_NATOCrate"] call A3A_fnc_log;
		};
	};
};

```

I also added the new debug logging to it. This drops the amount of code in it, some of the debug can be removed if required, does this still work in the way you expect it to?

### PR DESCRIPTION
**Pull request by Toakan**
## What type of PR is this.
1. [ ] Bug
2. [X] Enhancement

### What have you changed and why?
Information:
    I've changed the weighted function to use the built in one provided by the engine. There is also additional debugging for weapons and random magazine unlocks per weapon.

### Please specify which Issue this PR Resolves.
closes #485 

### Please verify the following and ensure all checks are completed.

1. [ ] Have you loaded the Mission in Singleplayer?
2. [X] Have you loaded the Mission in a Dedicated Server?

### Is further testing or are further changes required?
1. [ ] No
2. [X] Yes (Please provide further detail below.)

********************************************************
Notes:

Review and please let me know what you think.

### CODE REVIEW

**Comment by Spoffy** [comment]
The weight changes and magazine changes seem good, though it's still missing one of the loops from the old code, that dealt with the situation where a category might be unavailable - either having no items in it, or being effectively empty because everything was unlocked. 

It was causing crates to frequently spawn with less items. SMGs on vanilla (no dlc) is the most egregious example, because it only has two items in - both we start with, so it's effectively empty.

File: A3-Antistasi/functions/Ammunition/fn_NATOCrate.sqf [hunk 1]
```diff
@@ -31,34 +31,31 @@ if (typeOf _crate == vehNATOAmmoTruck) then {
...
-	private _weight = _x select 1;
-	
-	for "_i" from 1 to _weight do {
-		//Add the collection to the types array.
-		_weightedWeaponCategories pushBack (_x select 0);
-	};
-} forEach _lootWeaponCategoryWeighting;
+//This overrides the shotgun setting.
+if (hasRHS) then {
+	_weaponlootWeighting set [8, 1.8];
```
Comment (jaj22) [inline A3-Antistasi/functions/Ammunition/fn_NATOCrate.sqf]: Doesn't this line set the wrong array element? Should be 7 if it's zero-based.

Comment (Toakan) [reply A3-Antistasi/functions/Ammunition/fn_NATOCrate.sqf]: Yup, you're right on that one. Nice catch.

**Comment by Toakan** [comment]
I think that edgecase catch needs to be reviewed. Removing the array values, or setting them to '0' weight would be a better way to do this.

**Comment by jaj22** [comment]
I'm not clear on the intentions here. If the point of removing unlocked items is that every crate item is valuable, then putting substantial weight on shotguns, SMGs and pistols seems contrary. Those don't often appear in the faction class lists, so a realism justification doesn't fit either.

If the goal is to improve performance by avoiding array subtractions (eg. allRifles - unlockedRifles) then I suspect the 10x retries on easily-unlocked categories would remove the benefits.

Array subtractions would also permit easy re-weighting of categories (eg. setting empties to zero) after creating the weight array. If you really want to avoid the array subtractions, and don't want to maintain lockedWeapon lists, you could instead drop the retries and simply generate more items at higher war tiers to compensate for the proportion that will be thrown away due to matching the unlocked lists.

Either way, I'd cut down the default shotgun/SMG/pistol weights. MGs also seem high compared to battlefield commonality, although that's probably to combat the 10 marginally different types of M249. Sniper rifles will be a bit broken because the categorisation is a disaster in RHS: The M24, M2010, M40 and SVD are all "assault rifles", for example. I think you'd get much better results and save a lot of complex code by making hardcoded loot lists per template, but if the decision's been made then it's been made.

Generating magazines per weapon seems like a bad move for most reasons (generation performance, network load, display spam, arsenal fragmentation), but I'd support having some ammo variety. Maybe roll a gaussian random number of mags per stack, and then create half the mags as entry 0 and half as a random type.

**Comment by PoweredByPot** [comment]
@jaj22 the whole intention of the underlying change to loot generation was two-fold:
1) we have plans to modify the information provided in templates to allow a more fluid selection of faction/side for each "side" in the mission. The old loot generation system of scraping units loadouts for items meant that specific units classes with appropriate gear would always have to exist in the templates and that is counter to our goals with them.

2) Scraping unit loadouts for equipment was very taxing on the host and also left out a massive selection of equipment and weapons that players could utilize.

So, that being said we are now in the process of trying to balance the availability of loot that could appear in a given crate. One of the most apparent issues that led to us using weights in the spawning was that launchers ended up completely drowned out of the massive volume of other weapons available. In the interest of trying to keep scripts small and simple it was decided that a weighted array would be a better option than using separate logic for each class, or other various options.

Largely we are trying to move away from static lists and definitions in the hopes that we can become more flexible and playable across various maps and modsets _without_ requiring extensive dev work to make them all compatible.

**Comment by Bob-Murphy** [comment]
Just stating to piss of @PoweredByPot that I liked the completely random loot crates :)

Love ya, pot
3>

**Comment by Spoffy** [comment]
@jaj22 - Avoiding array subtracting is intended as a performance optimisation, yes.  For comparison with some real numbers:

`allWeapons` - 264 class names
`allRifles` - 214 class names
`allWeapons - allRifles` - 1.1100ms
`selectRandom allWeapons` - 0.0015 ms
`x in allWeapons` - Where X is the last item, so worst case - 0.0112ms

So even 10 random selections, with 10 allWeapons checks, comes to 0.127ms - Which is 10x faster than the array subtraction. I'm still going to look for better way to do it though.

In terms of weighting - that's still open for discussion. We could cut down SMG/Pistol/Shotgun weights without too much issue.

The primary goal of the loot crates is to be interesting, and to provide exotic (otherwise unattainable) weapons right now. That could in theory change later, but right now, that's their main purpose. 

Hardcoding lists is possible - but Pot's explained above why we've gone with the config scraping approach.

Magazines per weapon - that's still being iterated on, but we'll take the feedback into account! :)

**Comment by jaj22** [comment]
I suspect the 10 * count A / count B iterations calculation is upside down, although neither orientation is going to work well when B is larger than A (small category + large unlockedWeapons). Too few rerolls if the category isn't fully unlocked, and too many if it is.

I'm good with the principle of switching between array subtraction and rerolls depending on the array sizes, but given that you're already creating the category-specific unlockedXXX arrays, wouldn't it work much better if those were used as input B to `pickRandomFromAProbablyNotInB`? For example by defining _weaponLootWeighting like this:

```
private _weaponlootWeighting = [
	[allRifles, unlockedRifles], 3,
	[allHandguns, unlockedHandguns], 1.2,
	[allMachineGuns, unlockedMachineGuns], 2,
        ...
];
```

From there you could also throw in the fully-unlocked zero-weighting with a simple loop, which would obsolete the RHS shotguns check.

**Comment by Spoffy** [comment]
@jaj22 - You're absolutely right with the iterations calculation. It is the wrong way round, and it really does need to use allRifles and unlockedRifles, rather than allWeapons and unlockedWeapons - otherwise the maths doesn't work. The maths only works when allX is the maximum size unlockedX can reach - heuristic falls down otherwise. I'll sort it now. 

With regards to filtering out shotguns, I've added a separate step above that does that.

**Comment by Spoffy** [comment]
Merging this in - but please keep commenting on this if you spot any other issues, and I'll fix it in unstable!

**Pull request merged**